helix-term/src/handlers/diagnostics.rs

```diff
@@ -1,12 +1,157 @@
+use std::collections::HashMap;
+use std::path::PathBuf;
+use std::time::Duration;
+
+use helix_core::syntax::LanguageServerFeature;
 use helix_event::{register_hook, send_blocking};
+use helix_lsp::lsp::{self, Diagnostic};
+use helix_lsp::LanguageServerId;
 use helix_view::document::Mode;
-use helix_view::events::DiagnosticsDidChange;
+use helix_view::events::{DiagnosticsDidChange, DocumentDidChange};
 use helix_view::handlers::diagnostics::DiagnosticEvent;
+use helix_view::handlers::lsp::PullDiagnosticsEvent;
 use helix_view::handlers::Handlers;
+use helix_view::Editor;
+use tokio::time::Instant;
 
 use crate::events::OnModeSwitch;
+use crate::job;
+
+const TIMEOUT: u64 = 120;
+
+#[derive(Debug)]
+pub(super) struct PullDiagnosticsHandler {}
+
+impl PullDiagnosticsHandler {
+    pub fn new() -> PullDiagnosticsHandler {
+        PullDiagnosticsHandler {}
+    }
+}
+
+impl helix_event::AsyncHook for PullDiagnosticsHandler {
+    type Event = PullDiagnosticsEvent;
+
+    fn handle_event(
+        &mut self,
+        event: Self::Event,
+        _: Option<tokio::time::Instant>,
+    ) -> Option<tokio::time::Instant> {
+        match event {
+            PullDiagnosticsEvent::Trigger => {}
+        }
+        Some(Instant::now() + Duration::from_millis(TIMEOUT))
+    }
+
+    fn finish_debounce(&mut self) {
+        job::dispatch_blocking(move |editor, _| pull_diagnostic_for_current_doc(editor))
+    }
+}
+
+fn pull_diagnostic_for_current_doc(editor: &mut Editor) {
+    let (_, doc) = current!(editor);
+
+    for language_server in doc.language_servers_with_feature(LanguageServerFeature::PullDiagnostics)
+    {
```

Comment:
Could we check for a language server with a pull diagnostics feature in the hook first? That way we avoid dispatching a callback via `send_blocking` if the server doesn't support it (which will likely be the most common case)

Reply:
I added `has_language_server_with_feature` as suggested by @the-mikedavis. This way only send when there is at least one language server with this feature.